<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good) Exercises</span></div>

**Exercise 1**

In [7]:
listofsquared = []

for number in range(0, 10):

    if number % 2 == 0:
        continue
 
    else:
        number_squared = number**2
        listofsquared.append(number_squared)

print(listofsquared)

[1, 9, 25, 49, 81]


**Exercise 2**

In [17]:
#part 1 (what do you want to do with the dummy var) 
#part 2 for statement 
# part 3 if statement

[number**2 for number in range(10) if number % 2 == 1]

[1, 9, 25, 49, 81]

**Exercise 3**

In [3]:
%%timeit
listofsquared = []
for number in range(0, 10):

    if number % 2 == 0:
        continue
 
    else:
        number_squared = number**2
        listofsquared.append(number_squared)


697 ns ± 13 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [4]:
%%timeit
[number**2 for number in range(10) if number % 2 == 1]

673 ns ± 1.07 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


Using list comprehension is faster than using a loop due to its efficiency in iteration as compared to using loop, which must handle the overhead Python interpreter with each iteration. 

**Exercise 4**

In [4]:
# Initial quantity of the radioactive substance
initial_quantity = 100

# Decay rate per year
decay_rate = 0.05

# Target quantity (half of initial amount)
target_quantity = initial_quantity / 2

# Initialize current quantity to initial quantity
current_quantity = initial_quantity

# Initialize years taken
years_taken = 0

# Loop until current quantity is reduced to half or below
while current_quantity > target_quantity:
    # Calculate the quantity after decay for this year
    current_quantity = current_quantity * (1 - decay_rate)
    # Increase the years taken by 1
    years_taken = years_taken + 1

years_taken

14

**Exercise 5**

In [5]:
co2_data = [
    (2000, 369.55), (2001, 371.14), (2002, 373.28), 
    (2003, 375.80), (2004, 377.52), (2005, 379.80), 
    (2006, 381.90), (2007, 383.79), (2008, 385.60), 
    (2009, 387.43), (2010, 389.90), (2011, 391.65), 
    (2012, 393.85), (2013, 396.52), (2014, 398.65),
    (2015, 400.83), (2016, 404.24), (2017, 406.55), 
    (2018, 408.52), (2019, 411.44), (2020, 414.24)
]

In [27]:
# started the range from 1 because comparison needs to be done
# first element cannot be compared with anything
for i in range(1, len(co2_data)):

    # store previous year's CO2 levels
    year1_co2_level = co2_data[i-1][1] 
    
    # store current year's CO2 levels
    year2_co2_level = co2_data[i][1]
    
    # store current year
    year2 = co2_data[i][0]
    
    # store change between current and previous years' CO2 levels
    change = year2_co2_level-year1_co2_level
    
    if change >=3:
    
        print(year2, change)

2016 3.410000000000025


The year that showed an increase of CO2 levels of 3 ppm or more compared to the previous year is 2016, with a change in concentration of 3.41 ppm.